language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -347,20 +347,24 @@ genDefDataType curPkgId conName mod tpls def =
                      ["daml.STATIC_IMPLEMENTS_SERIALIZABLE_CHECK<" <> conName <> ">(" <> conName <> ")"]
             in ((typeDesc, serDesc), Set.unions $ map (Set.setOf typeModuleRef . snd) bs)
         DataEnum enumCons ->
-          let
-            typeDesc =
-                [ "export enum " <> conName <> " {"] ++
-                [ "  " <> cons <> " = " <> "\'" <> cons <> "\'" <> ","
-                | VariantConName cons <- enumCons] ++
-                [ "}"
-                , "daml.STATIC_IMPLEMENTS_SERIALIZABLE_CHECK<" <> conName <> ">(" <> conName <> ")"
-                ]
-            serDesc =
-                ["() => jtv.oneOf<" <> conName <> ">" <> "("] ++
-                ["  jtv.constant(" <> conName <> "." <> cons <> ")," | VariantConName cons <- enumCons] ++
-                [");"]
+          let cs = map unVariantConName enumCons
+              typeDesc =
+                let td = ["  " <> "'" <> cons <> "'" | cons <- cs]
+                    td' = reverse td
+                in [""] <> reverse (head td' : map (<> " | ") (tail td'))
+              -- The complete definition of the companion object.
+              serDesc =
+                [ "export const " <> conName <> ": daml.Serializable<" <> conName <> "> & {"] <>
```

Comment:
Side note: The whole code in this module seems to become increasingly messy. I hope it won’t grow much anymore so this is probably not a huge issue but if it does, we should use a pretty printing library and maybe consider introducing some intermediate types or at least lots of helper functions instead of converting directly to strings.

Reply:
Captured in issue https://github.com/digital-asset/daml/issues/5047.